### ISSUE TITLE
A11y strings with Talkback will be incorrectly formatted for many locales, when localized 

### ISSUE THREAD

**Issue by Delphine** [mozilla-mobile/fenix#27762]
I just exposed [this string](https://github.com/mozilla-mobile/fenix/blob/main/app/src/main/res/values/strings.xml#L1894) on Pontoon, to only later realize that this will cause issues down the road.

Indeed, the comments says `<!-- Action label for links that point to an article. Talkback will append this to say "Double tap to read the article". -->`, however the string reads `read the article`, which assumes that all languages have the same sentence structure, and that "Double tap to" necessarily comes before "read the article" (which is incorrect).

I then realized that there are already a few a11y strings supposed to work in conjunction with Talkback and that present the same issue, according to the [comments here](https://github.com/mozilla-mobile/fenix/blob/main/app/src/main/res/values/strings.xml#L1886). See the four strings part of this a11y service.

I don't know how to work around this, but wanted to make sure this is on the radar, and that we understand this won't work for many locales.
CC @cpeterso

**Comment by cpeterso** [comment]
Thanks for catching that. Gluing strings together is definitely not safe for L10n! I will raise this issue in Fenix sprint planning tomorrow.

**Comment by cpeterso** [comment]
@Delphine Do you know of any languages that these strings will cause L10n problems for Talkback?

@Mugurell tested Romanian and French and didn't see any problems with those languages.

**Comment by Delphine** [comment]
@cpeterso Thanks for asking! So both those languages that were tested are romance languages, and this will likely work for other similar languages.
But what about languages like Russian, Chinese, Arabic, Welsh, or more?
Unfortunately that's an assumption we can't make, and we ship 100+ locales on Firefox for Android :/ 
But also unfortunately, I don't know how to work around this specific case since it uses another service (Talkback)

**Comment by mcarare** [comment]
Looking at the example below, taken from the [documentation](https://developer.android.com/reference/android/view/accessibility/AccessibilityNodeInfo.AccessibilityAction#AccessibilityAction(int,%20java.lang.CharSequence)) for setting a custom label for an action to be interpreted by screen readers, we can see that it is expected that a custom label is localized in this manner:

```
AccessibilityAction action = new AccessibilityAction(
           AccessibilityAction.ACTION_CLICK.getId(), getLocalizedLabel());
```

Technical note: Looking further in the code we can see that the label param [changes](https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/view/accessibility/AccessibilityNodeInfo.java#5201) the value of `mValue` variable, that is later used to "describe" a node:

`return "AccessibilityAction: " + getActionSymbolicName(mActionId) + " - " + mLabel;`(https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/view/accessibility/AccessibilityNodeInfo.java#5257).

Explaining how Talkback later interpretes this info it's a bit more complicated and unnecessary for the purpose of this issue, it is sufficient to notice that **the label is always appended after the action symbolic name** (`ACTION_CLICK` in our case).
 
The documentation itself should be enough to clarify the localization aspect, but the code also enforces the idea that the current description for the click action is correctly localized by just translating the action label, without considering other sentence structures. 

**Additional note:** The above code handles adding a custom action label for views, AFAIK the Compose `SemanticsProperties` work in the same way in composing the node info for the screen readers to interpret.

The only thing that I would advise in using these labels is to keep them as short as possible, "native" ones being almost always as short as one word.

**Comment by cpeterso** [comment]
@mcarare Thanks for explaining how Talkback handles the strings. Sounds like Fenix is doing the right thing and don't need to change any code or strings. Should we close this issue?

**Comment by mcarare** [comment]
Yes! IMO the localization is done correctly and the comment explanation should be enough for localizers to have the proper context.

**Comment by cpeterso** [comment]
@Delphine, thanks for raising this issue. Based on @mcarare's investigation (and [this L10n code comment in Fenix](https://github.com/mozilla-mobile/fenix/blob/3c9ba2bbaa0168124cdc39b8c2871a4a65a43e5d/app/src/main/res/values/strings.xml#L1886)), I think we are localizing our a11y labels appropriately for the Talkback API.

**Comment by Delphine** [comment]
"“it is sufficient to notice that the label is always appended after the action symbolic name” "
But it's not... A lot of languages don't work like that. 
And it's not, if the language is supported in the client but not the OS (which is the case of many locales on Fenix)

**Comment by flodolo** [comment]
I can't claim to even be able to read Kotlin code, but what about the approach described in [this article](https://medium.com/microsoft-mobile-engineering/android-accessibility-resolving-common-talkback-issues-3c45076bcdf6) (search for `Action descriptions,` can't find how to link to the section)

**Comment by mcarare** [comment]
> "“it is sufficient to notice that the label is always appended after the action symbolic name” " But it's not... A lot of languages don't work like that. And it's not, if the language is supported in the client but not the OS (which is the case of many locales on Fenix)

@Delphine I am not talking about natural language sentence formation, but the way Talkback constructs the sentence, and it's the same order, regardless of the language.

**Comment by Delphine** [comment]
Ok thanks for clarifying and looking into all this @mcarare! It's too nad but it looks like there are no workarounds, unfortunately.
